src/fq.rs

```diff
@@ -57,28 +57,25 @@ const MODULUS: Fq = Fq([
     0x73eda753299d7d48,
 ]);
 
-/// Compute a + b + carry, returning the result and setting carry to the
-/// carry value.
+/// Compute a + b + carry, returning the result the new carry over.
 #[inline(always)]
-fn adc(a: u64, b: u64, carry: &mut u128) -> u64 {
-    *carry = u128::from(a) + u128::from(b) + (*carry >> 64);
-    *carry as u64
+fn adc(a: u64, b: u64, carry: u64) -> (u64, u64) {
+    let adc = u128::from(a) + u128::from(b) + u128::from(carry);
+    (adc as u64, (adc >> 64) as u64)
 }
 
-/// Compute a - (b + borrow), returning the result and setting borrow to
-/// the borrow value.
+/// Compute a - (b + carry), returning the result and the carry over
 #[inline(always)]
-fn sbb(a: u64, b: u64, borrow: &mut u128) -> u64 {
-    *borrow = u128::from(a).wrapping_sub(u128::from(b) + (*borrow >> 127));
-    *borrow as u64
+fn sbb(a: u64, b: u64, carry: u64) -> (u64, u64) {
```

Comment:
The code should be this, I think:

```rust
fn sbb(a: u64, b: u64, borrow: u64) -> (u64, u64) {
    let tmp = u128::from(a).wrapping_sub(u128::from(b) | u128::from(borrow >> 63));
    (tmp as u64, (tmp >> 64) as u64)
}
```

1. We definitely still want to refer to this as a `borrow` value since that's what it is
2. We should use `tmp` inside these functions, since `let sbb = ...` is rebinding `sbb` from the function definition which probably violates some kind of Rust coding convention.
3. We want to OR in the borrow value of the previous invocation rather than adding like I was doing before.

This will compile to a pretty optimal:

```
        shrq    $63, %rdx
        orq     %rsi, %rdx
        xorl    %ecx, %ecx
        subq    %rdx, %rdi
        sbbq    $0, %rcx
```

Reply:
1. Ahh... I guess I assumed borrow was because we were borrowing, and that the term was carry under.
2. I'll use `ret` to a similar effect.
3. Using that snippet as is causes some of the tests to fail.